pkg/object/function/worker/ingress.go

```diff
@@ -0,0 +1,362 @@
+		superSpec *supervisor.Spec
+
+		faasNetworkLayerURL string
+		faasHostSuffix      string
+		faasNamespace       string
+
+		namespace string
+		mutex     sync.RWMutex
+
+		tc             *trafficcontroller.TrafficController
+		pipelines      map[string]struct{}
+		httpServer     *supervisor.ObjectEntity
+		httpServerSpec *supervisor.Spec
+	}
+
+	pipelineSpecBuilder struct {
+		Kind              string `yaml:"kind"`
+		Name              string `yaml:"name"`
+		httppipeline.Spec `yaml:",inline"`
+	}
+)
+
+// newIngressServer creates an initialized ingress server
+func newIngressServer(superSpec *supervisor.Spec, super *supervisor.Supervisor,
+	controllerName string) *ingressServer {
+	entity, exists := super.GetSystemController(trafficcontroller.Kind)
+
+	if !exists {
+		panic(fmt.Errorf("BUG: traffic controller not found"))
+	}
+
+	tc, ok := entity.Instance().(*trafficcontroller.TrafficController)
+	if !ok {
+		panic(fmt.Errorf("BUG: want *TrafficController, got %T", entity.Instance()))
+	}
+	return &ingressServer{
+		pipelines:  make(map[string]struct{}),
+		httpServer: nil,
+		super:      super,
+		superSpec:  superSpec,
+		mutex:      sync.RWMutex{},
+		namespace:  fmt.Sprintf("%s/%s", superSpec.Name(), "ingress"),
+		tc:         tc,
+	}
+}
+
+func newPipelineSpecBuilder(funcName string) *pipelineSpecBuilder {
+	return &pipelineSpecBuilder{
+		Kind: httppipeline.Kind,
+		Name: funcName,
+		Spec: httppipeline.Spec{},
+	}
+}
+
+func (b *pipelineSpecBuilder) yamlConfig() string {
+	buff, err := yaml.Marshal(b)
+	if err != nil {
+		logger.Errorf("BUG: marshal %#v to yaml failed: %v", b, err)
+	}
+	return string(buff)
+}
+
+func (b *pipelineSpecBuilder) appendReqAdaptor(funcSpec *spec.Spec, faasNamespace, faasHostSuffix string) *pipelineSpecBuilder {
+	adaptorName := "requestAdaptor"
+	b.Flow = append(b.Flow, httppipeline.Flow{Filter: adaptorName})
+
+	b.Filters = append(b.Filters, map[string]interface{}{
+		"kind":   requestadaptor.Kind,
+		"name":   adaptorName,
+		"method": funcSpec.RequestAdaptor.Method,
+		"path":   funcSpec.RequestAdaptor.Path,
+		"header": funcSpec.RequestAdaptor.Header,
+
+		// let faas Provider's gateway recognized this function by Host field
+		"host": funcSpec.Name + "." + faasNamespace + "." + faasHostSuffix,
+	})
+
+	return b
+}
+
+func (b *pipelineSpecBuilder) appendProxy(faasNetworkLayerURL string) *pipelineSpecBuilder {
+	mainServers := []*proxy.Server{
+		{
+			URL: faasNetworkLayerURL,
+		},
+	}
+
+	backendName := "faasBackend"
+
+	lb := &proxy.LoadBalance{
+		Policy: proxy.PolicyRoundRobin,
+	}
+
+	b.Flow = append(b.Flow, httppipeline.Flow{Filter: backendName})
+	b.Filters = append(b.Filters, map[string]interface{}{
+		"kind": proxy.Kind,
+		"name": backendName,
+		"mainPool": &proxy.PoolSpec{
+			Servers:     mainServers,
+			LoadBalance: lb,
+		},
+	})
+
+	return b
+}
+
+func (ings *ingressServer) httpServerYAML(httpServer *httpserver.Spec) string {
+	ingressHTTPServerFormat := `
```

Comment:
Propose to use struct, it could probe change of it while compiling.

Reply:
Sure. 
Updated it with adding `httpServerSpecBuilder` already.